src/install/npm.zig

```diff
@@ -824,15 +824,24 @@ pub const PackageManifest = struct {
         {
             const releases = this.pkg.releases.keys.get(this.versions);
             var i = releases.len;
+
+            var bestMatchVersion: ?Semver.Version = null;
+            var bestMatchPackage: ?PackageVersion = null;
             // For now, this is the dumb way
             while (i > 0) : (i -= 1) {
                 const version = releases[i - 1];
-                const packages = this.pkg.releases.values.get(this.package_versions);
 
-                if (group.satisfies(version)) {
-                    return .{ .version = version, .package = &packages[i - 1] };
+                // If we find one that matches, save it, but keep looping because we might find a newer match.
+                // The versions from the registry are not in any particular order.
+                if (group.satisfies(version) and (bestMatchVersion == null or Semver.Version.gt(version, bestMatchVersion.?))) {
+                    bestMatchVersion = version;
+                    bestMatchPackage = this.pkg.releases.values.get(this.package_versions)[i - 1];
                 }
             }
+
+            if (bestMatchVersion != null and bestMatchPackage != null) {
+                return .{ .version = bestMatchVersion.?, .package = &bestMatchPackage.? };
```

Comment:
`&bestMatchPackage.?` is a pointer to stack memory, which is invalid and will eventually cause a segfault. This is a very common mistake to make for people knew to Zig (and something Zig should make easier, IMO)

Related to:
- https://github.com/ziglang/zig/issues/5725 
- https://github.com/ziglang/zig/issues/2646 
- https://github.com/ziglang/zig/issues/2301

Reply:
That makes sense, but does that mean the existing code is broken too?

> return .{ .version = version, .package = &packages[i - 1] };

* https://github.com/oven-sh/bun/blob/main/src/install/npm.zig#L833
* https://github.com/oven-sh/bun/blob/main/src/install/npm.zig#L846

Reply:
Those pointers are not allocated on the stack, they are allocated on the heap. Their lifetime is tied to the lifetime of the `packages` array, which is longer than the current stack (the function call)

Reply:
This for example is valid:
```zig
var packages = bun.default_allocator.alloc(u32, 100);
return &packages[1];
```

This is invalid:
```zig
var packages = [_]u32{0} ** 100;
return &packages[1];
```

Reply:
ooooh, the problem isn't the pointer, it's the local variable!

Sorry, I haven't used pointers since college in the late 90's!